src/jit/lower.cpp

```diff
@@ -338,24 +338,96 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
     case GT_CAST:
         {
             assert(tree->gtPrev == tree->gtGetOp1());
-            NYI_IF(tree->gtOverflow(), "TYP_LONG cast with overflow");
-            switch (tree->AsCast()->CastFromType())
+            GenTree* srcOp = tree->gtGetOp1();
+            var_types srcType = tree->AsCast()->CastFromType();
+            var_types dstType = tree->AsCast()->CastToType();
+
+            if (tree->gtFlags & GTF_UNSIGNED)
             {
-            case TYP_INT:
-                if (tree->gtFlags & GTF_UNSIGNED)
+                srcType = genUnsignedType(srcType);
+            }
+
+            if (genActualTypeIsIntOrI(srcType) || varTypeIsGC(srcType))
```

Comment:
I think this could/should be simply `varTypeIsIntegralOrI(srcType)`. I just noticed that both `varTypeIsIntOrI()` and `genActualTypeIsIntOrI()` don't actually include types with the `VTF_I` classification, which seems at a minimum quite confusing.